mhkit/dolfyn/rotate/base.py

```diff
@@ -0,0 +1,299 @@
+
+       - heading is defined as the direction the x-axis points, positive
+         clockwise from North (this is *opposite* the right-hand-rule
+         around the Z-axis), range 0-360 degrees.
+
+       - pitch is positive when the x-axis pitches up (this is *opposite* the
+         right-hand-rule around the Y-axis)
+
+       - roll is positive according to the right-hand-rule around the
+         instrument's x-axis
+
+    """
+    if units.lower() == 'degrees':
+        pitch = np.deg2rad(pitch)
+        roll = np.deg2rad(roll)
+        heading = np.deg2rad(heading)
+    elif units.lower() == 'radians':
+        pass
+    else:
+        raise Exception("Invalid units")
+        
+    # Converts the DOLfYN-defined heading to one that follows the right-hand-rule
+    # reports heading as rotation of the y-axis positive counterclockwise from North
+    heading = np.pi / 2 - heading 
+                                  
+    # Converts the DOLfYN-defined pitch to one that follows the right-hand-rule.
+    pitch = -pitch
+
+    ch = np.cos(heading)
+    sh = np.sin(heading)
+    cp = np.cos(pitch)
+    sp = np.sin(pitch)
+    cr = np.cos(roll)
+    sr = np.sin(roll)
+    zero = np.zeros_like(sr)
+    one = np.ones_like(sr)
+
+    H = np.array(
+        [[ch, sh, zero],
+         [-sh, ch, zero],
+         [zero, zero, one], ])
+    P = np.array(
+        [[cp, zero, -sp],
+         [zero, one, zero],
+         [sp, zero, cp], ])
+    R = np.array(
+        [[one, zero, zero],
+         [zero, cr, sr],
+         [zero, -sr, cr], ])
+
+    return np.einsum('ij...,jk...,kl...->il...', R, P, H)
+
+
+def orient2euler(omat):
+    """
+    Calculate DOLfYN-defined euler angles from the orientation matrix.
+
+    Parameters
+    ----------
+    omat : |np.ndarray|
+      The orientation matrix
+
+    Returns
+    -------
+    heading : |np.ndarray|
+      The heading angle. Heading is defined as the direction the x-axis points,
+      positive clockwise from North (this is *opposite* the right-hand-rule
+      around the Z-axis), range 0-360 degrees.
+    pitch : np.ndarray
+      The pitch angle (degrees). Pitch is positive when the x-axis 
+      pitches up (this is *opposite* the right-hand-rule around the Y-axis).
+    roll : np.ndarray
+      The roll angle (degrees). Roll is positive according to the 
+      right-hand-rule around the instrument's x-axis.
+
+    """
+
+    if isinstance(omat, np.ndarray) and \
+            omat.shape[:2] == (3, 3):
+        pass
+    elif hasattr(omat, 'orientmat'):
+        omat = omat['orientmat'].values
+        
+    # Note: orientation matrix is earth->inst unless supplied by an external IMU
+    hh = np.rad2deg(np.arctan2(omat[0, 0], omat[0, 1]))
+    hh %= 360
+    return (
+        # heading 
+        hh,
+        # pitch 
+        np.rad2deg(np.arcsin(omat[0, 2])),
+        # roll
+        np.rad2deg(np.arctan2(omat[1, 2], omat[2, 2])),
+    )
+
+
+def q2orient(quaternions):
```

Comment:
spell out so that its more clear? ie quaternion2orient

Reply:
This can be done. I was aiming for shorter function names